### ISSUE TITLE
Segfault in readbuffer_gets()

### ISSUE THREAD

**Issue by rkrejci** [openssl/openssl#23915]
Hi,
we are using libcurl in our application and we had a (one-time, we are not able to reproduce) segfault in Solaris. Here is a part of the coredump's stack listing:
```
fd7fae58 libcrypto.so.3`BIO_test_flags+4(0?, f?, 0?, feabc000?, ffd5f3d8?, 49b440?)
fd7faeb8 libcrypto.so.3`BIO_copy_next_retry+8(59c6e8?, 4f0348?, 1?, 0?, ffffff?, 0?)
fd7faf18 libcrypto.so.3`readbuffer_gets+0x14c(59c6e8?, 442690?, ffffffff?, 59c698?, 4f0348?, 0?)
fd7faf78 libcrypto.so.3`BIO_gets+0xa0(59c6e8?, 442690?, ff?, 442798?, 601?, 0?)
fd7fafe8 libcrypto.so.3`ossl_prov_bio_gets+0x30(47e6e0?, 442690?, ff?, 1?, ffffff?, 0?)
fd7fb048 libcrypto.so.3`BIO_gets+0xa0(535fe0?, 442690?, ff?, feabc000?, 1?, 0?)
fd7fb0b8 libcrypto.so.3`PEM_read_bio_ex+0xc8(535fe0?, 2?, fd7fb1c4?, fd7fb1c8?, 442690?, 2?)
fd7fb148 libcrypto.so.3`pem2der_decode+0x40(fd7fb1cc?, 47e6e0?, 535fe0?, fe1bb100?, fd7fb2b4?, fe22a1c0?)
fd7fb240 libcrypto.so.3`decoder_process+0x534(5745f0?, fd7fb33c?, 3c7ee0?, 0?, 5745f0?, 17?)
fd7fb2c8 libcrypto.so.3`OSSL_DECODER_from_bio+0x7c(5745f0?, 59c6e8?, 100c0102?, 57460c?, 1?, 5745f0?)
fd7fb350 libcrypto.so.3`pem_read_bio_key+0x138(59c6e8?, 0?, 0?, 0?, 0?, 0?)
fd7fb3e0 libcrypto.so.3`PEM_read_bio_PrivateKey_ex+0x20(, ...)
fd7fb448 libcrypto.so.3`PEM_read_bio_RSAPrivateKey+0x10(0?, 0?, 0?, 0?, 0?, fe2e1e80?)
fd7fb4a8 sslctx_function+0xc8(0?, 44d018?, 0?, 1?, 0?, 0?)
fd7fb518 libcurl.so.4.8.0`ossl_connect_step1+0x50c(0?, 4be8d0?, 4f4318?, 1?, 41226c?, 4bec54?)
```

I traced it down into openssl and I believe that I found a bug. There is a code pattern around `BIO_read()` used at several places:
```
    if (b->next_bio == NULL)
        return 0;
    ...
    i = BIO_read(b->next_bio, data, data_len);
    if (i <= 0) {
        BIO_copy_next_retry(b);
        ...
```
But in the case of [`readbuffer_gets()`](https://github.com/openssl/openssl/blob/dc9bc6c8e1bd329ead703417a2235ab3e97557ec/crypto/bio/bf_readbuff.c#L218), the pattern misses the NULL pointer check at the beginning. So in the case `b->next_bio == NULL`, the `BIO_read()` returns `-1` and `BIO_copy_next_retry()` segfaults because of accessing `b->next_bio->flags`.

In our case, the segfault appeared with openssl 3.0.13, but I see the same issue also in the current master branch.

I can prepare a PR, but wanted to discuss the issue here beforehand in case the missing check is intended here for some reason. But I don't see a way a caller could cause an invalid/unexpected BIO structure here.

**Comment by nhorman** [comment]
yeah, that looks legitimate to me.  It sort of begs the question as to why next_bio wasn't set in that instance, but adding the NULL check makes sense.  Please submit your PR when ready, and thank you!

**Comment by t8m** [comment]
I am not saying #23918 should not be merged however it is not really fixing the true issue. I.e., why the readbuffer_gets() was called with next_bio == NULL. This really should not happen and might be something more serious.

**Comment by rkrejci** [comment]
I agree, but unfortunately, we are not able to reproduce the segfault in this case and since I'm not an expert for openssl internals (I wasn't even 100% sure that `next_bio == NULL` is something that should never happen), I wasn't able to trace it anywhere else to get know the cause of having NULL there.

**Comment by rkrejci** [comment]
Hi, I'm still waiting for the decision regarding CCLA, but I have some update regarding the root cause of the issue.

Here is a simple code to reproduce the issue (tested on Linux with OpenSSL 3.1.4):
```
#include <openssl/ssl.h>

int main(int argc, char *argv[])
{
    RSA *rsa = NULL;

    rsa = PEM_read_bio_RSAPrivateKey(NULL, NULL, 0, NULL);

    return 0;
}
```

1. I know that `PEM_read_bio_RSAPrivateKey()` is deprecated and I'm going to update our code
2. I know that the call (arguments) does not make sense and I'm going to handle it in our code, but I expect that libcrypto is still not supposed to crash in this case. 

The question here is if the provided patch is good enough for this use case, this situation may be probably detected earlier. I'm still not sure about the `next_bio` pointer and when it is supposed to be set - it is not done by `BIO_new_ex()` (called by `pem_read_bio_key()` when the given BIO is NULL), so in the case here it goes down with the initial NULL.

```
rkrejci@localhost:~/workspace/tmp> LD_PRELOAD=./libcrypto.so.3 valgrind --num-callers=50 ./test
==8477== Memcheck, a memory error detector
==8477== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==8477== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
==8477== Command: ./test
==8477== 
==8477== Invalid read of size 4
==8477==    at 0x49568CB: BIO_test_flags (bio_lib.c:213)
==8477==    by 0x4957EB4: BIO_copy_next_retry (bio_lib.c:816)
==8477==    by 0x495439B: readbuffer_gets (bf_readbuff.c:275)
==8477==    by 0x495737C: BIO_gets (bio_lib.c:480)
==8477==    by 0x4964122: ossl_core_bio_gets (ossl_core_bio.c:108)
==8477==    by 0x4C521F0: ossl_prov_bio_gets (bio_prov.c:110)
==8477==    by 0x4C524BB: bio_core_gets (bio_prov.c:183)
==8477==    by 0x495737C: BIO_gets (bio_lib.c:480)
==8477==    by 0x4B93CB5: get_name (pem_lib.c:760)
==8477==    by 0x4B94369: PEM_read_bio_ex (pem_lib.c:945)
==8477==    by 0x4B94697: PEM_read_bio (pem_lib.c:1006)
==8477==    by 0x4C7BB97: read_pem (decode_pem2der.c:40)
==8477==    by 0x4C7BD2D: pem2der_decode (decode_pem2der.c:143)
==8477==    by 0x4A4C04B: decoder_process (decoder_lib.c:971)
==8477==    by 0x4A4A540: OSSL_DECODER_from_bio (decoder_lib.c:82)
==8477==    by 0x4B95635: pem_read_bio_key_decoder (pem_pkey.c:60)
==8477==    by 0x4B95DE0: pem_read_bio_key (pem_pkey.c:241)
==8477==    by 0x4B96086: PEM_read_bio_PrivateKey_ex (pem_pkey.c:304)
==8477==    by 0x4B960C8: PEM_read_bio_PrivateKey (pem_pkey.c:312)
==8477==    by 0x4B8FC55: PEM_read_bio_RSAPrivateKey (pem_all.c:73)
==8477==    by 0x401175: main (test.c:6)
==8477==  Address 0x30 is not stack'd, malloc'd or (recently) free'd
==8477== 
==8477== 
==8477== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==8477==  Access not within mapped region at address 0x30
==8477==    at 0x49568CB: BIO_test_flags (bio_lib.c:213)
==8477==    by 0x4957EB4: BIO_copy_next_retry (bio_lib.c:816)
==8477==    by 0x495439B: readbuffer_gets (bf_readbuff.c:275)
==8477==    by 0x495737C: BIO_gets (bio_lib.c:480)
==8477==    by 0x4964122: ossl_core_bio_gets (ossl_core_bio.c:108)
==8477==    by 0x4C521F0: ossl_prov_bio_gets (bio_prov.c:110)
==8477==    by 0x4C524BB: bio_core_gets (bio_prov.c:183)
==8477==    by 0x495737C: BIO_gets (bio_lib.c:480)
==8477==    by 0x4B93CB5: get_name (pem_lib.c:760)
==8477==    by 0x4B94369: PEM_read_bio_ex (pem_lib.c:945)
==8477==    by 0x4B94697: PEM_read_bio (pem_lib.c:1006)
==8477==    by 0x4C7BB97: read_pem (decode_pem2der.c:40)
==8477==    by 0x4C7BD2D: pem2der_decode (decode_pem2der.c:143)
==8477==    by 0x4A4C04B: decoder_process (decoder_lib.c:971)
==8477==    by 0x4A4A540: OSSL_DECODER_from_bio (decoder_lib.c:82)
==8477==    by 0x4B95635: pem_read_bio_key_decoder (pem_pkey.c:60)
==8477==    by 0x4B95DE0: pem_read_bio_key (pem_pkey.c:241)
==8477==    by 0x4B96086: PEM_read_bio_PrivateKey_ex (pem_pkey.c:304)
==8477==    by 0x4B960C8: PEM_read_bio_PrivateKey (pem_pkey.c:312)
==8477==    by 0x4B8FC55: PEM_read_bio_RSAPrivateKey (pem_all.c:73)
==8477==    by 0x401175: main (test.c:6)
==8477==  If you believe this happened as a result of a stack
==8477==  overflow in your program's main thread (unlikely but
==8477==  possible), you can try to increase the size of the
==8477==  main thread stack using the --main-stacksize= flag.
==8477==  The main thread stack size used in this run was 8388608.
==8477== 
==8477== HEAP SUMMARY:
==8477==     in use at exit: 121,247 bytes in 3,407 blocks
==8477==   total heap usage: 4,099 allocs, 692 frees, 150,322 bytes allocated
==8477== 
==8477== LEAK SUMMARY:
==8477==    definitely lost: 0 bytes in 0 blocks
==8477==    indirectly lost: 0 bytes in 0 blocks
==8477==      possibly lost: 0 bytes in 0 blocks
==8477==    still reachable: 121,247 bytes in 3,407 blocks
==8477==         suppressed: 0 bytes in 0 blocks
==8477== Rerun with --leak-check=full to see details of leaked memory
==8477== 
==8477== For lists of detected and suppressed errors, rerun with: -s
==8477== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
Segmentation fault (core dumped)
```
The fixed code returns NULL from `PEM_read_bio_RSAPrivateKey()`:
```
rkrejci@localhost:~/workspace/tmp> LD_PRELOAD=./libcrypto.so.3.fix valgrind ./test
==8047== Memcheck, a memory error detector
==8047== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==8047== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
==8047== Command: ./test
==8047== 
==8047== 
==8047== HEAP SUMMARY:
==8047==     in use at exit: 0 bytes in 0 blocks
==8047==   total heap usage: 4,117 allocs, 4,117 frees, 152,327 bytes allocated
==8047== 
==8047== All heap blocks were freed -- no leaks are possible
==8047== 
==8047== For lists of detected and suppressed errors, rerun with: -s
==8047== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)

```

**Comment by nhorman** [comment]
Given that the cla for the PR is not yet submitted, I'm moving this to inactive for now (to be closed at the end of the 3.4 dev cycle).  If you have gotten your ccla signed please update the issue here
